### ISSUE TITLE
Support NotInList in dynamic finders

### ISSUE THREAD

**Issue by sdelamo** [grails/grails-core#9718]
Today I stumbled upon the next problem:

```
def countPersonNotAnoymous = Person.countByIdNotInList(anonymousPersonIds)
log.error("count with CountBy: ${countPersonNotAnoymous}")

countPersonNotAnoymous = Person.createCriteria().get {
    projections {
        countDistinct "id"
    }
    not {'in'("id",anonymousPersonIds)}
}
log.error("count with Criteria: ${countPersonNotAnoymous}")
```

anonymousPersonIds is a list of ids from a Domain class. 

I expected both log statements to return the same output however they did not. 

I turned on sql logging and the query generated by `Person.countByIdNotInList` is:

`Hibernate: select count(*) as y0_ from person this_ where this_.id in (?)`

Note the above SQL query does not include a not statement. 

I am running Grails 3.0.14

**Comment by graemerocher** [comment]
The expression `NotInList` is not a supported expression right now. See

http://grails.github.io/grails-doc/latest/guide/GORM.html#finders

Section "Method Expressions" for the list of supported expressions.

**Comment by droggo** [comment]
Shouldn't this fail if not supported method name is used? Silently ignoring seems dangerous here.

**Comment by jeffbrown** [comment]
@droggo Of course.

I don't think the fix is to have it fail though.  I think we should just support it.

**Comment by graemerocher** [comment]
making it fail is not possible. You would break all property names that are equals to `idNotInList`, maybe something fairly unlikely, but there are a lot of potential finder combinations that could be broken if you become more strict about validating what and what isn't allowed

**Comment by sdelamo** [comment]
A failure would have helped me most but I understand your concerns.

**Comment by jasonschwab** [comment]
Just got bit by this as well, didn't check the docs assuming the IDE gave it to me as an option while making the dynamic finder and took me a bit to figure out Foo.findAllByIdInList([1]) was returning the exact same resultset as Foo.findAllByIdNotInList([1])

**Comment by graemerocher** [comment]
This seems more of an IDE issue then, since `NotInList` has never been supported in the entire history of Grails

**Comment by macrosak** [comment]
@graemerocher as far as I can see `NotInList` works as expected in Grails 2.2.4

Also this documentation mentions `Not` as a valid operator: http://docs.grails.org/3.2.3/ref/Domain%20Classes/findAllBy.html

**Comment by snimavat** [comment]
@graemerocher cant conclude from comments what is done here.
Is this implemented, is it going to through error, or its unchanges and "wont fix" ?

**Comment by graemerocher** [comment]
It has been implemented in 6.1 see https://github.com/grails/grails-data-mapping/issues/828

**Comment by snimavat** [comment]
Ok, thanks for clarifying
